create QueryAction and Action when a card has is_write set

escherize opened the pull request:
implements #22905
Part of https://github.com/metabase/metabase/issues/22542

This will generate an Action and QueryAction entry when:

1) a card is created with `is_write` = `true`
2) a card is updated s.t. `is_write` is changed to `true`

And delete an Action and QueryAction entry when a card is updated s.t. `is_write` is changed to `false`

escherize committed:
create QueryAction and Action when a card has is_write set

escherize committed:
Some fixups 🧱

escherize committed:
include action in models to exclude

escherize committed:
delete {query_}actions when is_write is set false

escherize committed:
linter fixes 🪡

escherize committed:
reorder models to exclude

escherize committed:
Merge remote-tracking branch 'origin/pod-writeback' into create-action-and-qa-when-is_write

escherize committed:
add more models to exclude

snoe commented:
Resolves https://github.com/metabase/metabase/issues/22905

camsaul commented on src/metabase/models/card.clj:
```diff
@@ -208,6 +209,19 @@
                                  (describe-database query-db-id)))
                           {:status-code 400})))))))
 
+(defn- create-actions-when-is-writable [{is-write? :is_write card-id :id}]
```
```suggestion
(defn- create-actions-when-is-writable! [{is-write? :is_write card-id :id}]
```

camsaul commented on src/metabase/models/card.clj:
```diff
@@ -208,6 +209,19 @@
+      (db/insert! action/QueryAction {:card_id card-id
+                                      :action_id action-id}))))
+
+(defn- delete-actions-when-not-writable [{is-write? :is_write card-id :id}]
```
```suggestion
(defn- delete-actions-when-not-writable! [{is-write? :is_write card-id :id}]
```

camsaul commented on test/metabase/models/card_test.clj:
```diff
@@ -291,3 +291,40 @@
+(deftest action-creation-test
+  (testing "actions are created when is_write is set"
+    (testing "during create"
+      (let [{card-id :id} (db/insert! Card (assoc (tt/with-temp-defaults Card) :is_write true))
+            {:keys [action_id card_id] :as qa-rows} (db/select-one QueryAction :card_id card-id)]
```
You should use `with-temp` here which will delete the Cards after the test. Otherwise if you run these tests from the REPL you will end up with a ton of new garbage Cards in your application DB each time you run the test. Generally the tests are supposed to clean up after themselves. Another option is `with-model-cleanup` if you still want to be able to use `insert!` manually instead of having `with-temp` do it for you. `with-temp` ultimately calls `insert!` tho so you can use it for this stuff

escherize committed:
respond to review comments

snoe committed:
Mysql is unhappy with nested queries in DELETE

Split the queries into a select and delete, and only run the delete if
there's something to remove which will be rare.

snoe committed:
Mysql is unhappy with alias in delete

snoe committed:
Merge remote-tracking branch 'origin/pod-writeback' into create-action-and-qa-when-is_write

snoe committed:
Merge branch 'pod-writeback' into create-action-and-qa-when-is_write
